File: cpp/0200_Number_of_Islands.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

class Solution {
public:
    void dfs(vector<vector<char>>& grid, int x, int y) {
        vector<vector<int>> dir = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

        grid[x][y] = '0';

        for (auto d : dir) {
            int nx = x + d[0];
            int ny = y + d[1];

            if (nx < 0 || ny < 0 || nx >= grid.size() || ny >= grid[0].size() || grid[nx][ny] != '1') continue;

            dfs(grid, nx, ny);
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        int ret = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == '1') {
                    dfs(grid, i, j);
                    ret++;
                }
            }
        }

        return ret;
    }
};

#if 0
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int ret = 0;
        queue<vector<int>> q;
        vector<vector<int>> dir = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == '1') {
                    ret++;

                    q.push({i, j});
                    while (!q.empty()) {
                        int size = q.size();
                        for (int k = 0; k < size; k++) {
                            int x = q.front()[0];
                            int y = q.front()[1];
                            q.pop();

                            grid[x][y] = '0';
                            for (auto d : dir) {
                                int nx = x + d[0];
                                int ny = y + d[1];

                                if (nx < 0 || ny < 0 || nx >= grid.size() || ny >= grid[0].size() || grid[nx][ny] != '1') continue;

                                q.push({nx, ny});
                            }
                        }
                    }
                }
            }
        }

        return ret;
    }
};
#endif

int main(int argc, char *argv[]) {
    Solution solution;
    vector<vector<char>> grid;

    // Input: grid = [
    // ["1","1","1","1","0"],
    // ["1","1","0","1","0"],
    // ["1","1","0","0","0"],
    // ["0","0","0","0","0"]
    // ]
    // Output: 1
    grid = {};
    grid.push_back({'1', '1', '1', '1', '0'});
    grid.push_back({'1', '1', '0', '1', '0'});
    grid.push_back({'1', '1', '0', '0', '0'});
    grid.push_back({'0', '0', '0', '0', '0'});
    cout << solution.numIslands(grid) << endl;

    // Input: grid = [
    // ["1","1","0","0","0"],
    // ["1","1","0","0","0"],
    // ["0","0","1","0","0"],
    // ["0","0","0","1","1"]
    // ]
    // Output: 3
    grid = {};
    grid.push_back({'1', '1', '0', '0', '0'});
    grid.push_back({'1', '1', '0', '0', '0'});
    grid.push_back({'0', '0', '1', '0', '0'});
    grid.push_back({'0', '0', '0', '1', '1'});
    cout << solution.numIslands(grid) << endl;

    return 0;
}

File: py/0200_Number_of_Islands.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


class Solution:
    def dfs(self, i: int, j: int, grid: List[List[str]]):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] != "1":
            return

        grid[i][j] = "0"
        delta = ((0, -1), (1, 0), (0, 1), (-1, 0))
        for dx, dy in delta:
            self.dfs(i + dx, j + dy, grid)

    def numIslands(self, grid: List[List[str]]) -> int:
        ret = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    ret += 1
                    self.dfs(i, j, grid)
        return ret

        # BFS
        """
        ret = 0
        queue = deque()
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    ret += 1
                    queue.append((i, j))
                    while len(queue) != 0:
                        x, y = queue.popleft()
                        grid[x][y] = 0
                        delta = ((0, -1), (1, 0), (0, 1), (-1, 0))
                        for dx, dy in delta:
                            nx, ny = x + dx, y + dy
                            if nx >= 0 and nx < len(grid) and ny >= 0 and ny < len(grid[0]) and grid[nx][ny] == "1":
                                queue.append((nx, ny))
        return ret
        """


if __name__ == "__main__":
    solution = Solution()

    grid = [["1", "1", "1", "1", "0"], ["1", "1", "0", "1", "0"], ["1", "1", "0", "0", "0"], ["0", "0", "0", "0", "0"]]
    print(solution.numIslands(grid))

    sys.exit(0)
